RecoLocalCalo/HcalRecAlgos/src/DoMahiAlgo.cc

```diff
@@ -0,0 +1,563 @@
+      updateWork_ = aTbVec_ - aTaMat_*ampVec_;
+      
+      Index idxwmaxprev = idxwmax;
+      double wmaxprev = wmax;
+      wmax = updateWork_.tail(nActive).maxCoeff(&idxwmax);
+      
+      if (wmax<threshold || (idxwmax==idxwmaxprev && wmax==wmaxprev)) {
+	break;
+      }
+      
+      if (iter>=nMaxItersNNLS_) {
+	break;
+      }
+
+      //unconstrain parameter
+      Index idxp = nP_ + idxwmax;
+      NNLSUnconstrainParameter(idxp);
+
+    }
+
+    while (true) {
+      if (nP_==0) break;     
+
+      ampvecpermtest_ = ampVec_;
+      
+      eigen_solve_submatrix(aTaMat_,aTbVec_,ampvecpermtest_,nP_);
+
+      //check solution
+      bool positive = true;
+      for (unsigned int i = 0; i < nP_; ++i)
+        positive &= (ampvecpermtest_(i) > 0);
+      if (positive) {
+        ampVec_.head(nP_) = ampvecpermtest_.head(nP_);
+        break;
+      } 
+
+      //update parameter vector
+      Index minratioidx=0;
+      
+      // no realizable optimization here (because it autovectorizes!)
+      double minratio = std::numeric_limits<double>::max();
+      for (unsigned int ipulse=0; ipulse<nP_; ++ipulse) {
+	if (ampvecpermtest_.coeff(ipulse)<=0.) {
+	  const double c_ampvec = ampVec_.coeff(ipulse);
+	  const double ratio = c_ampvec/(c_ampvec-ampvecpermtest_.coeff(ipulse));
+	  if (ratio<minratio) {
+	    minratio = ratio;
+	    minratioidx = ipulse;
+	  }
+	}
+      }
+      ampVec_.head(nP_) += minratio*(ampvecpermtest_.head(nP_) - ampVec_.head(nP_));
+      
+      //avoid numerical problems with later ==0. check
+      ampVec_.coeffRef(minratioidx) = 0.;
+      
+      NNLSConstrainParameter(minratioidx);
+    }
+   
+    ++iter;
+
+    //adaptive convergence threshold to avoid infinite loops but still
+    //ensure best value is used
+    if (iter%10==0) {
+      threshold *= 10.;
+    }
+    
+    break;
+  }
+  
+  return true;
+}
+
+bool DoMahiAlgo::OnePulseMinimize() {
+
+  invcovp_ = covDecomp_.matrixL().solve(pulseMat_);
+
+  SingleMatrix aTamatval = invcovp_.transpose()*invcovp_;
+  SingleVector aTbvecval = invcovp_.transpose()*covDecomp_.matrixL().solve(amplitudes_);
+
+  ampVec_.coeffRef(0) = std::max(0., aTbvecval.coeff(0)/aTamatval.coeff(0));
+
+  return true;
+
+}
+
+double DoMahiAlgo::CalculateChiSq() {
+
+  return (covDecomp_.matrixL().solve(pulseMat_*ampVec_ - amplitudes_)).squaredNorm();
+}
+
+void DoMahiAlgo::setPulseShapeTemplate(const HcalPulseShapes::Shape& ps) {
+
+  if (!(&ps == currentPulseShape_ ))
+    {
+      resetPulseShapeTemplate(ps);
+      currentPulseShape_ = &ps;
+    }
+}
+
+double DoMahiAlgo::getSiPMDarkCurrent(double darkCurrent, double fcByPE, double lambda) {
+  double mu = darkCurrent * 25 / fcByPE;
+  return sqrt(mu/pow(1-lambda,3)) * fcByPE;
+}
+
+void DoMahiAlgo::resetPulseShapeTemplate(const HcalPulseShapes::Shape& ps) { 
+  ++ cntsetPulseShape_;
+  psfPtr_.reset(new FitterFuncs::PulseShapeFunctor(ps,false,false,false,false,1,0,2.5,0,0.00065,1,10));
```

Comment:
please document the magic values (inline comments may be enough if these are known from first principles; otherwise, some configuration parameters may be needed)